### PR TITLE
Combine Activity Panel count requests into one endpoint

### LINKED ISSUE #60393: Number of wp-json calls for wc-analytics and wc-admin
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

A user opens /wp-admin/edit.php?post_type=product&paged=7

This triggers 6 wp-json requests from woocommerce alone:
1) /wp-json/wc-analytics/orders?page=1&per_page=1&status%5B0%5D=processing&status%5B1%5D=on-hold&_fields%5B0%5D=id&_locale=user
2) /wp-json/wc-analytics/products/reviews?page=1&per_page=1&status=hold&_embed=1&_fields%5B0%5D=id&_locale=user
3) /wp-json/wc-analytics/products/count-low-in-stock?status=publish&page=1&per_page=1&_locale=user
4) /wp-json/wc-analytics/orders?page=1&per_page=1&status%5B0%5D=processing&status%5B1%5D=on-hold&_fields%5B0%5D=id&_locale=user
5) /wp-json/wc-analytics/products/reviews?page=1&per_page=1&status=hold&_embed=1&_fields%5B0%5D=id&_locale=user
6) /wp-json/wc-analytics/products/count-low-in-stock?status=publish&page=1&per_page=1&_locale=user

Then there are some other plugins that trigger 1 or 2 or an admin-ajax.php call, and opening a single admin page triggers 15 requests in addition to the main request.

This is obviously not an issue that comes up in testing or with most stores, because they're too small, but even for a small company with only 10 people using wp-admin simultaneously during peak hours, this mean's that you essentially have to spin up a single server/instance with at least 75 php-fpm workers, to handle the up to 10*15 (practically they don't hit all at the same time, which is why you get away with 50% of max requests) requests.

The wp-admin page is obviously not page cached, so there is no reason why this data cannot be output directly on the actual document request, making all these wp-json calls redundant.

### Expected behavior

Directly output the data you need on the main doc request

### Actual behavior

Sends 100 wp-json subrequests

### Steps to reproduce

Open e.g. /wp-admin/edit.php?post_type=product&paged=7 (or any other page nr) which has various products showing

### WordPress Environment

WP 6.8.2
WC 10.0

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by faisalahammad**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Opening any wp-admin page fires 6 separate `wc-analytics` REST requests just to populate the Activity Panel (header bell icon plus the homescreen "Things to do next" panel): 3 distinct endpoints (orders to fulfill, reviews to moderate, low stock count), each called independently by 2 different components, so none of them share a cached resolution.

This adds a single `GET /wc-analytics/activity-panel/counts` endpoint that returns all three counts in one response (delegating internally to the existing endpoints, so no counting logic is duplicated), plus a shared `activityPanelStore` selector in `@woocommerce/data`. All three Activity Panel components now read from this one selector, so the resolution cache collapses them to a single network request per page load. The three original endpoints are untouched, this is additive only.

Closes #60393.

### Screenshots or screen recordings:

No UI changes. This is a performance optimization that reduces 6 REST API requests to 1 per page load. The Activity Panel badge and counts display identically to before.

### How to test the changes in this Pull Request:

1. Open `/wp-admin/edit.php?post_type=product` with the browser Network tab open, filtered to `wc-analytics`. Confirm the Activity Panel triggers exactly one request, to `/wc-analytics/activity-panel/counts`, instead of the previous 6 requests to `orders`, `products/reviews`, and `products/count-low-in-stock`.
2. Create an order in `processing` or `on-hold` status, an unapproved product review, and a product at or below its low stock threshold.
3. Reload wp-admin and open the header bell icon. Confirm "Orders to fulfill", "Reviews to moderate", and "Inventory to review" show correct counts. Open the WooCommerce Home screen and confirm the "Things to do next" panel shows the same counts.
4. As a user without `manage_woocommerce` (e.g. Subscriber), confirm a direct request to `/wc-analytics/activity-panel/counts` returns 403.

### Testing that has already taken place:

- Added `ActivityPanelCountsTest.php` covering response shape, permission check, custom `order_statuses` param, and count accuracy against the existing low-stock endpoint. All passing.
- Added Jest tests for the new store's reducer, selectors, and resolver. All passing.
- Manually verified the endpoint end-to-end against a running wp-env instance (returns 200 with the expected shape).
- Ran PHPCS, PHPStan, and `tsc --build` on all changed files, all clean.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

- [ ] Automatically create a changelog entry from the details below.
- [x] This Pull Request does not require a changelog entry. (Comment required below)

Changelog Entry Details

#### Significance

- [ ] Patch
- [x] Minor
- [ ] Major

#### Type

- [ ] Fix - Fixes an existing bug
- [ ] Add - Adds functionality
- [ ] Update - Update existing functionality
- [ ] Dev - Development related task
- [ ] Tweak - A minor adjustment to the codebase
- [x] Performance - Address performance issues
- [ ] Enhancement - Improvement to existing functionality

#### Message

Combine the Activity Panel's orders, reviews, and low stock count requests into a single wc-analytics/activity-panel/counts endpoint.

Changelog Entry Comment

#### Comment

Created manually (`plugins/woocommerce/changelog/fix-60393-combine-activity-panel-count-requests` and `packages/js/data/changelog/add-60393-activity-panel-store`).

### Release Communication

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [ ] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

### CODE REVIEW

File: plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		}
+
+		$headers = $response->get_headers();
+		if ( isset( $headers['X-WP-Total'] ) ) {
+			return (int) $headers['X-WP-Total'];
+		}
+
+		$data = $response->get_data();
+		return isset( $data['total'] ) ? (int) $data['total'] : 0;
+	}
```
[the remark this answers is not in the document]
Comment (faisalahammad) [reply plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php:121]: Fixed. `get_count_via()` returns `null` instead of `0` on sub-request failure, schema fields marked `nullable`, and TS type widened to `number | null`. Added a test forcing a sub-request failure via `rest_pre_dispatch` to confirm the field comes back null while the other counts stay correct. See 917398d.

**Comment by faisalahammad** [comment]
Re: CodeRabbit's nitpick on centralizing the `counts?.field ?? default` mapping (not flagged actionable) — skipping this one on purpose. The 3 call sites use different fallback semantics: `homescreen/activity-panel/index.js` falls back to `null` (unknown/unresolved, matching sibling `getOrdersTotalCount(..., 0)`-style handling), while `activity-panel.js` and `abbreviated-notifications-panel.js` fall back to `0` (safe value for a `> 0` check). A shared selector would have to pick one convention and lose that distinction, so I left the 3 sites as-is.

Pushed 917398d for the actionable finding: `get_count_via()` now returns `null` instead of `0` on sub-request failure, schema fields marked nullable, TS type widened, and added a regression test.

File: packages/js/data/src/activity-panel/resolvers.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * low stock products) in a single request instead of one per count.
+ */
+export function* getActivityPanelCounts() {
+	try {
+		const counts: ActivityPanelCounts = yield apiFetch( {
+			path: `${ NAMESPACE }/activity-panel/counts`,
+		} );
+		yield getActivityPanelCountsSuccess( counts );
+	} catch ( error ) {
+		yield getActivityPanelCountsError( error );
```
Comment (chihsuan) [inline packages/js/data/src/activity-panel/resolvers.ts:27]: Since the catch resolves normally, the resolver won't retry, and nothing reads `getActivityPanelCountsError` so a failed request just shows 0 in the bell (looks dentical to "all clear"), and now takes all three counts down together instead of one. Could we surface the error or rethrow so a retry can kick in? 🙏

File: plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+						'_fields'  => array( 'id' ),
+					)
+				),
+				'reviews_to_moderate_count'   => $this->get_count_via(
+					'/wc-analytics/products/reviews',
+					array(
+						'page'     => 1,
+						'per_page' => 1,
+						'status'   => $request->get_param( 'review_status' ),
+						'_embed'   => 1,
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php]: nit: we only read X-WP-Total, so the embed is wasted work here. Safe to drop.

File: plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	 */
+	private function get_count_via( $route, $params ) {
+		$sub_request = new \WP_REST_Request( 'GET', $route );
+		foreach ( $params as $key => $value ) {
+			$sub_request->set_param( $key, $value );
+		}
+
+		$response = rest_do_request( $sub_request );
+
+		if ( is_wp_error( $response ) || $response->is_error() ) {
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php]: nit: `rest_do_request()` never returns a `WP_Error` (it wraps errors into the response), so `is_error()` alone covers it.

File: plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'activity_panel_counts',
+			'type'       => 'object',
+			'properties' => array(
+				'orders_to_fulfill_count'     => array(
+					'description' => __( 'Number of orders to fulfill. Null if the underlying sub-request failed.', 'woocommerce' ),
+					'type'        => 'integer',
+					'context'     => array( 'view' ),
+					'readonly'    => true,
+					'nullable'    => true,
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/API/ActivityPanelCounts.php]: Two small ones: nullable isn't enforced by WP (use `'type' => array( 'integer', 'null' )`), and since `get_counts()` returns a raw array without `prepare_item_for_response()`, the `schema/context/_fields` aren't actually applied. Might be worth trimming what doesn't take effect.

**Review by chihsuan — COMMENTED** [review]
Thanks for contributing! @faisalahammad Nice improvement. 👍 

Two things I'd want sorted before merge:


1. **Stale counts after in-panel actions** — the badge count moved to `activityPanelStore`, but the existing invalidations still refresh the old stores (`reviews/index.js:373`, `stock/index.js:51`), so approving a review or updating stock no longer updates the badge until reload. Needs an `invalidateResolutionForStoreSelector('getActivityPanelCounts')` in those flows.
2. **Silent failure** — a failed combined request now renders `0` in the bell with no retry, and takes all three counts down together.

 Both trace back to the same root (no invalidation + swallowed error), so they can likely be fixed together. Rest are perf ideas / nits inline

**Comment by faisalahammad** [comment]
@chihsuan Thanks for the review. All four points addressed in babaf80:

**1. Stale counts after in-panel actions** — Added `invalidateResolution( 'getActivityPanelCounts', [] )` on `activityPanelStore` in both `reviews/index.js` (clearReviewsCache) and `stock/index.js` (updateStock). Badge now refreshes after approve/delete/stock-update without reload.

**2. Silent error in resolver** — Removed try/catch from `resolvers.ts`. Error now propagates, `@wordpress/data` marks resolution failed and can retry. Updated Jest test to match.

**3. Remove `_embed`** — Dropped `'_embed' => 1` from reviews sub-request params. Only `X-WP-Total` is read, embed was wasted work.

**4. `is_wp_error()` dead check** — Removed `is_wp_error( $response ) ||`. `rest_do_request()` wraps errors into the response, so `is_error()` alone covers it.

**5. Schema nullable type + dead fields** — Changed each field from `'type' => 'integer', 'nullable' => true` to `'type' => array( 'integer', 'null' )`. Removed `context` and `readonly` since `get_counts()` returns a raw array without `prepare_item_for_response()`.

**6. Final fallback** — Changed `return isset( $data['total'] ) ? (int) $data['total'] : 0` to `... : null` to match the error-path semantics.

File: packages/js/data/src/activity-panel/resolvers.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * Request the Activity Panel counts (orders to fulfill, reviews to moderate,
+ * low stock products) in a single request instead of one per count.
+ */
+export function* getActivityPanelCounts() {
+	const counts: ActivityPanelCounts = yield apiFetch( {
+		path: `${ NAMESPACE }/activity-panel/counts`,
+	} );
+	yield getActivityPanelCountsSuccess( counts );
```
Comment (chihsuan) [inline packages/js/data/src/activity-panel/resolvers.ts]: Sorry, I noticed that you removed the `try/catch` in babaf80b65f8d7b5a1902c0d9298d24af6863f3f. I think we should actually keep them. Can you undo it? Thanks!

**Review by chihsuan — COMMENTED** [review]
@faisalahammad 👋  CI is red on this branch: `Lint`, `JavaScript [unit]` (admin-library), `PHP 8.5 [unit:php]`, and `Blocks e2e 2/10` are all failing. Could you take a
  look? 

Also left one code note inline.

File: packages/js/data/src/activity-panel/resolvers.ts (hunk 6 above)
Comment (faisalahammad) [reply packages/js/data/src/activity-panel/resolvers.ts]: Restored the try/catch in 88c92e8. Added back the error dispatch test case too (throw into the generator, expect getActivityPanelCountsError).

**Comment by faisalahammad** [comment]
@chihsuan Pushed 88c92e8 fixing CI:

- Lint: added missing `declare( strict_types=1 );` to `ActivityPanelCounts.php` and its test, fixed equals-sign alignment on test line 140.
- JS unit: `stock/test/index.js` was still using the old dispatch mock, missing `invalidateActivityPanel`. Added the mock and asserted it gets called with `getActivityPanelCounts`.
- Restored the resolver `try/catch` per your request, with the error-dispatch test case back too.

The PHP unit timeout (shard 2/4) and the Blocks e2e cart-store nonce failure looked unrelated to this diff (infra timeout and a Store API nonce flake), so no code change there — should get a fresh run on the new push.

**Comment by chihsuan** [comment]
Thanks @faisalahammad It seems that CI jobs fail during dependency install. I think the current CI failures are not caused by your changes.

This branch is ~20 commits behind trunk. Could you sync with trunk to refresh the lockfile? Thanks.

**Comment by faisalahammad** [comment]
Pushed `60460fa` addressing the CodeRabbit nitpick: renamed `stock_status` param to `product_status` (it controls post status, not WC stock status). Updated PR description with Screenshots section (no UI changes noted).

Ready for re-review.

**Comment by chihsuan** [comment]
Hey @faisalahammad, could you please help resolve the branch conflicts when you get a chance? thank you for your patience.

**Comment by faisalahammad** [comment]
Hi @chihsuan, branch conflicts resolved by merging upstream/trunk. One conflict in reviews/index.js resolved cleanly. Ready for re-review.

**Comment by faisalahammad** [comment]
Hi @chihsuan, I've cleaned up the branch history — the previous merge commit had pulled in some unrelated files (.claudeignore, etc.). Rebasing onto the latest `upstream/trunk` resolved this. The PR now shows only the 26 relevant files with 767 additions.

Conflict in `reviews/index.js` resolved the same way: keep the conditional invalidation plus `invalidateActivityPanel` call.

Ready for re-review.

**Review by chihsuan — APPROVED** [review]
